template/template.go

```diff
@@ -0,0 +1,197 @@
+package template
+
+import (
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+
+	ast "go.expect.digital/mf2/parse"
+)
+
+type execFn func(operand any, opts map[string]any) (string, error)
+
+var execFuncs = map[string]execFn{}
+
+var mutex sync.Mutex
+
+// AddFunc adds a function to the template's function map.
+func AddFunc(name string, f func(any, map[string]any) (string, error)) {
+	mutex.Lock()
+	execFuncs[name] = f
+	mutex.Unlock()
+}
+
+type Template ast.AST
+
+func New() *Template {
+	return new(Template)
+}
+
+func Must(t *Template, err error) *Template {
+	if err != nil {
+		panic(err)
+	}
+
+	return t
+}
+
+func (t *Template) Parse(input string) (*Template, error) {
+	ast, err := ast.Parse(input)
+	if err != nil {
+		return nil, syntaxErr(err)
+	}
+
+	*t = Template(ast)
+
+	return t, nil
+}
+
+func (t *Template) Execute(wr io.Writer, input map[string]any) error {
+	var resolve func() (string, error)
+
+	switch message := t.Message.(type) {
+	case nil:
+		return nil
+	case ast.SimpleMessage:
+		resolve = func() (string, error) { return resolveSimpleMessage(message, input) }
+	case ast.ComplexMessage:
+		return fmt.Errorf("'%T' not implemented", message) // TODO: Implement.
+	}
+
+	s, err := resolve()
+	if err != nil {
+		return fmt.Errorf("resolve message: %w", err)
+	}
+
+	if _, err = wr.Write([]byte(s)); err != nil {
```

Comment:
We should use `io.Writer` more efficiently. Pass it down to all functions that write string values and write to it directly.